File: src/emby_range_cache_proxy/ranges.py
from __future__ import annotations

import re

from .models import ByteRange

_RANGE_RE = re.compile(r"bytes=(\d*)-(\d*)$")


def parse_range_header(value: str | None, *, size: int) -> ByteRange:
    if size <= 0:
        raise ValueError("size must be positive")
    if not value:
        return ByteRange(0, size - 1)
    if "," in value:
        raise ValueError("multiple ranges are not supported")
    match = _RANGE_RE.fullmatch(value.strip())
    if not match:
        raise ValueError("invalid range header")
    left, right = match.groups()
    if left == "" and right == "":
        raise ValueError("empty range")
    if left == "":
        length = int(right)
        if length <= 0:
            raise ValueError("invalid suffix range")
        return ByteRange(max(0, size - length), size - 1)
    start = int(left)
    if start >= size:
        raise ValueError("range start beyond size")
    end = int(right) if right else size - 1
    if end < start:
        raise ValueError("range end before start")
    return ByteRange(start, min(end, size - 1))


def plan_playback_range(
    value: str | None,
    *,
    size: int,
    head_bytes: int,
    tail_bytes: int,
    default_open_range_bytes: int,
) -> ByteRange:
    byte_range = parse_range_header(value, size=size)
    if not value:
        return byte_range

    match = _RANGE_RE.fullmatch(value.strip())
    if match is None:
        return byte_range
    left, right = match.groups()
    if left == "" or right != "":
        return byte_range

    start = byte_range.start
    tail_start = max(0, size - tail_bytes)
    if start < min(head_bytes, size):
        return ByteRange(start, min(head_bytes, size) - 1)
    if start >= tail_start:
        return ByteRange(start, size - 1)
    return ByteRange(start, min(start + default_open_range_bytes - 1, size - 1))


def intersect_ranges(left: ByteRange, right: ByteRange) -> ByteRange | None:
    start = max(left.start, right.start)
    end = min(left.end, right.end)
    if end < start:
        return None
    return ByteRange(start, end)


def content_range_header(byte_range: ByteRange, *, size: int) -> str:
    return f"bytes {byte_range.start}-{byte_range.end}/{size}"

File: go/internal/ranges/ranges.go
package ranges

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/xmm2022/emby-range-cache-proxy/go/internal/model"
)

const (
	mib int64 = 1024 * 1024
)

var rangeRE = regexp.MustCompile(`^bytes=(\d*)-(\d*)$`)

func AdaptiveHeadTail(size int64) (int64, int64) {
	return 8 * mib, 8 * mib
}

func ParseRangeHeader(value string, size int64) (model.ByteRange, error) {
	if size <= 0 {
		return model.ByteRange{}, fmt.Errorf("size must be positive")
	}
	if value == "" {
		return model.ByteRange{Start: 0, End: size - 1}, nil
	}
	if strings.Contains(value, ",") {
		return model.ByteRange{}, fmt.Errorf("multiple ranges are not supported")
	}
	match := rangeRE.FindStringSubmatch(strings.TrimSpace(value))
	if match == nil {
		return model.ByteRange{}, fmt.Errorf("invalid range header")
	}
	left, right := match[1], match[2]
	if left == "" && right == "" {
		return model.ByteRange{}, fmt.Errorf("empty range")
	}
	if left == "" {
		length, err := strconv.ParseInt(right, 10, 64)
		if err != nil || length <= 0 {
			return model.ByteRange{}, fmt.Errorf("invalid suffix range")
		}
		start := size - length
		if start < 0 {
			start = 0
		}
		return model.ByteRange{Start: start, End: size - 1}, nil
	}
	start, err := strconv.ParseInt(left, 10, 64)
	if err != nil {
		return model.ByteRange{}, fmt.Errorf("invalid range start")
	}
	if start >= size {
		return model.ByteRange{}, fmt.Errorf("range start beyond size")
	}
	end := size - 1
	if right != "" {
		parsedEnd, err := strconv.ParseInt(right, 10, 64)
		if err != nil {
			return model.ByteRange{}, fmt.Errorf("invalid range end")
		}
		end = parsedEnd
	}
	if end < start {
		return model.ByteRange{}, fmt.Errorf("range end before start")
	}
	if end >= size {
		end = size - 1
	}
	return model.ByteRange{Start: start, End: end}, nil
}

func PlanPlaybackRange(value string, size, headBytes, tailBytes, defaultOpenRangeBytes int64, openHeadResponseBytes *int64) (model.ByteRange, error) {
	byteRange, err := ParseRangeHeader(value, size)
	if err != nil {
		return model.ByteRange{}, err
	}
	if value == "" {
		return byteRange, nil
	}
	match := rangeRE.FindStringSubmatch(strings.TrimSpace(value))
	if match == nil {
		return byteRange, nil
	}
	left, right := match[1], match[2]
	if left == "" || right != "" {
		return byteRange, nil
	}

	start := byteRange.Start
	tailStart := size - tailBytes
	if tailStart < 0 {
		tailStart = 0
	}
	headEnd := headBytes
	if headEnd > size {
		headEnd = size
	}
	if start < headEnd {
		end := headEnd - 1
		if openHeadResponseBytes != nil && *openHeadResponseBytes > 0 {
			capped := start + *openHeadResponseBytes - 1
			if capped < end {
				end = capped
			}
		}
		return model.ByteRange{Start: start, End: end}, nil
	}
	if start >= tailStart {
		return model.ByteRange{Start: start, End: size - 1}, nil
	}
	end := start + defaultOpenRangeBytes - 1
	if end >= size {
		end = size - 1
	}
	return model.ByteRange{Start: start, End: end}, nil
}

func ContentRangeHeader(byteRange model.ByteRange, size int64) string {
	return fmt.Sprintf("bytes %d-%d/%d", byteRange.Start, byteRange.End, size)
}
